### PR TITLE
Support for flutter run/build module on iOS

### LINKED ISSUE #19114: Support `flutter run` on a project of type module on iOS
Similar to https://github.com/flutter/flutter/issues/18997, just for iOS

### PR DESCRIPTION
**Pull request by mravn-google**
Makes `flutter run` and `flutter build` work for Flutter module projects on iOS.

Fixes https://github.com/flutter/flutter/issues/19114.

* Added template files for Flutter module to define an ephemeral host app for iOS.
* Added template "overlay" and logic to use CocoaPods only when Flutter project declares plugin dependencies.
* Expanded use of `IosProject` in favor of a folder path string.
* Made more getters of `XxxProject` synchronous.
* Code paths have generally not been prepared for materialization of the ephemeral host app. That will be done in an upcoming PR.
* A devicelab integration test will also be done in an upcoming PR (similar to `module_test` for Android).

### CODE REVIEW

**Comment by mravn-google** [comment]
cc @matthew-carroll @jamesderlin

File: packages/flutter_tools/lib/src/base/file_system.dart [hunk 1]
```diff
@@ -145,3 +146,16 @@ String canonicalizePath(String path) => fs.path.normalize(fs.path.absolute(path)
...
+///
+/// Returns true, if [file] does not exist.
+///
+/// Returns false, if [file] exists, but [referenceFile] does not.
+bool isOlderThanReference({@required File file, @required File referenceFile}) {
+  if (!file.existsSync())
+    return true;
+  return referenceFile.existsSync()
+      && referenceFile.lastModifiedSync().isAfter(file.lastModifiedSync());
+}
```
Comment (sigurdm) [inline packages/flutter_tools/lib/src/base/file_system.dart]: newline

File: packages/flutter_tools/lib/src/project.dart [hunk 2]
```diff
@@ -186,20 +208,46 @@ class IosProject {
     }
   }
 
+  void regenerateFromTemplate() {
```
Comment (sigurdm) [inline packages/flutter_tools/lib/src/project.dart]: could inline _shouldRegenerateFromTemplate here?

File: packages/flutter_tools/lib/src/project.dart [hunk 3]
```diff
@@ -186,20 +208,46 @@ class IosProject {
...
+
   Future<void> materialize() async {
     throwToolExit('flutter materialize has not yet been implemented for iOS');
   }
 
   bool _shouldRegenerateFromTemplate() {
-    return Cache.instance.fileOlderThanToolsStamp(directory.childFile('podhelper.rb'));
+    if (!isModule)
+      return false;
+    final File templateFile = directory.childDirectory('Flutter').childFile('podhelper.rb');
```
Comment (sigurdm) [inline packages/flutter_tools/lib/src/project.dart]: Use the directory modified timestamp

File: packages/flutter_tools/templates/module/README.md [hunk 4]
```diff
@@ -59,9 +59,31 @@ under app author control) Android host app with a dependency on the
...
+suitable only when the Flutter part declares no plugin dependencies.
+
+#### host_app_ephemeral_cocoapods
+
+Written to `.ios/` on top of `host_app_ephemeral`.
+ 
+Adds CocoaPods.
+
+Combined contents define an ephemeral host app suitable for when the
+Flutter part sdeclares plugin dependencies.
```
Comment (sigurdm) [inline packages/flutter_tools/templates/module/README.md]: s

File: packages/flutter_tools/templates/module/README.md [hunk 5]
```diff
@@ -59,9 +59,31 @@ under app author control) Android host app with a dependency on the
...
+`.ios/Flutter` folder contents.
+
+The host app does not make use of CocoaPods, and is therefore
+suitable only when the Flutter part declares no plugin dependencies.
+
+#### host_app_ephemeral_cocoapods
+
+Written to `.ios/` on top of `host_app_ephemeral`.
+ 
+Adds CocoaPods.
```
Comment (sigurdm) [inline packages/flutter_tools/templates/module/README.md]: ... support

File: packages/flutter_tools/templates/module/ios/host_app_ephemeral_cocoapods/Podfile.copy.tmpl [hunk 6]
```diff
@@ -0,0 +1,6 @@
+platform :ios, '8.0'
+
+target 'Runner' do
+  flutter_application_path = '../'
+  eval(File.read(File.join(flutter_application_path, '.ios', 'Flutter', 'podhelper.rb')))
+end
```
Comment (sigurdm) [inline packages/flutter_tools/templates/module/ios/host_app_ephemeral_cocoapods/Podfile.copy.tmpl]: Do we still need to disable bitcode for included third-party pods?

**Review by sigurdm — APPROVED** [review]
LGTM

File: packages/flutter_tools/lib/src/base/file_system.dart (hunk 1 above)
Comment (mravn-google) [reply packages/flutter_tools/lib/src/base/file_system.dart]: Done.

File: packages/flutter_tools/lib/src/project.dart (hunk 2 above)
Comment (mravn-google) [reply packages/flutter_tools/lib/src/project.dart]: Done.

File: packages/flutter_tools/lib/src/project.dart (hunk 3 above)
Comment (mravn-google) [reply packages/flutter_tools/lib/src/project.dart]: Done.

File: packages/flutter_tools/templates/module/README.md (hunk 5 above)
Comment (mravn-google) [reply packages/flutter_tools/templates/module/README.md]: Done.

File: packages/flutter_tools/templates/module/README.md (hunk 4 above)
Comment (mravn-google) [reply packages/flutter_tools/templates/module/README.md]: Done.

File: packages/flutter_tools/templates/module/ios/host_app_ephemeral_cocoapods/Podfile.copy.tmpl (hunk 6 above)
Comment (mravn-google) [reply packages/flutter_tools/templates/module/ios/host_app_ephemeral_cocoapods/Podfile.copy.tmpl]: It seems not.

File: packages/flutter_tools/lib/src/project.dart (hunk 2 above)
Comment (mravn-google) [reply packages/flutter_tools/lib/src/project.dart]: Done.

File: packages/flutter_tools/lib/src/project.dart (hunk 3 above)
Comment (mravn-google) [reply packages/flutter_tools/lib/src/project.dart]: Done.

File: packages/flutter_tools/templates/module/README.md (hunk 5 above)
Comment (mravn-google) [reply packages/flutter_tools/templates/module/README.md]: Done.

File: packages/flutter_tools/templates/module/README.md (hunk 4 above)
Comment (mravn-google) [reply packages/flutter_tools/templates/module/README.md]: Done.

**Review by mravn-google — COMMENTED** [review]
Thanks for the review.

**Comment by matthew-carroll** [comment]
CC @jamesderlin @xster @cbracken @chinmaygarde @gspencergoog 

Sorry for the spam, but all the iOS eyes we can get on these changes, the better.

**Pull request merged**